sklearn/neural_network/_rbm.py

```diff
@@ -370,7 +370,10 @@ def score_samples(self, X):
         ind = (np.arange(v.shape[0]), rng.randint(0, v.shape[1], v.shape[0]))
         if sp.issparse(v):
             data = -2 * v[ind] + 1
-            v_ = v + sp.csr_matrix((data.A.ravel(), ind), shape=v.shape)
+            if isinstance(data, np.matrix):
+                v_ = v + sp.csr_matrix((data.A.ravel(), ind), shape=v.shape)
+            else:
+                v_ = v + sp.csr_matrix((data.ravel(), ind), shape=v.shape)
```

Comment:
We need to document the surprises that come with this code. First we need to document the `np.matrix` vs. `np.array` due to the indexing.

Then, I would prefer to make it explicit which type of sparse thingy we have such that we add two sparse containers that are exactly of the same type. Let's imagine that in the future, someone permute `v` and the sparse matrix, then it will not provide the same container. Let's play safe.

```suggestion
            if sp.isspmatrix_csr(v):
                # `data` is a NumPy matrix with the previous indexing
                v_ = v + sp.csr_matrix((data.A.ravel(), ind), shape=v.shape)
            else:  # sp.csr_array
                # `data` is a NumPy array with the previous indexing
                v_ = v + sp.csr_array((data.ravel(), ind), shape=v.shape)
```